src/agents/graph.py

```diff
@@ -81,40 +109,47 @@ def _plan(self, state: AgentState) -> dict[str, Any]:
             ]
         ).partial(
             members=", ".join(self.members),
-            output_format=self.parser.get_format_instructions(),
+            output_format=self.plan_parser.get_format_instructions(),
         )
-        planner = planner_prompt | self.model.llm
+        planner_chain = planner_prompt | self.model.llm
         try:
-            plan = planner.invoke(
-                {
-                    "messages": state.messages[-1:],
+            plan_response = planner_chain.invoke(
+                input={
+                    "messages": filter_messages(state.messages),
                 },  # last message is the user query
-                config={"callbacks": [handler]},
             )
-            state.subtasks = self.parser.parse(plan.content).subtasks
-            return {
-                "subtasks": state.subtasks,
-                "messages": (
-                    [
-                        AIMessage(
-                            content=f"Task decomposed into subtasks and assigned to agents: {state.subtasks}",
-                            name=PLANNER,
-                        )
-                    ]
+
+            try:
+                plan = self.plan_parser.parse(plan_response.content)
+                if plan.response:
+                    return create_node_output(
+                        message=AIMessage(content=plan.response, name=PLANNER),
+                        final_response=plan.response,
+                        next=EXIT,
```

Comment:
This is the case, where the planner does not come up with a plan but instead has a direct response. 

Is the `next=EXIT` correct? Shouldn't this direct response go to the finalizer first?

Reply:
at the moment, we don't need another LLM call as planner already is able to answer the question, which makes it quite fast.
Forwarding it to finalizer creates another LLM call and sometimes creates fuzzy answer.

Reply:
- The way I understood it, all responses should go through the `Finalizer`. Is this not correct?
 - The `Planner` either returns a plan or a direct response. In case of a plan, next comes an agent. In case of a direct response, next comes `Exit`. Why does the generated graph show that the `Planner` can also go to the `Finalizer`?

Reply:
- If planner already can answer (e.g. general query) or error happens, it is immediately ending with exit node w/o finalizer.
- in case of plan, the plan is forwarded to supervisor where the agents called and eventually finalizer finalizes the messages of the agents. Finalizer is special for Kyma and K8s.